### PR TITLE
Add RFC: Enhanced Artifact Management in Agent Protocol

### PR DESCRIPTION
**Pull request by Swiftyos**
I've added a new RFC titled "Enhanced Artifact Management in Agent Protocol". This proposal is designed to address some limitations in our current artifact management system within the agent protocol. Please see [this issue](https://github.com/e2b-dev/agent-protocol/issues/30) @ntindle raised

## RFC Summary:

The newly proposed RFC aims to introduce more flexibility and clarity to our artifact handling. This includes the ability for agents to specify distinct file upload paths, effectively supporting external storage solutions such as AWS S3. Additionally, a distinction between artifacts generated by the agent and input artifacts will be introduced. Lastly, the RFC also touches upon revisiting and potentially revising the current artifact handling mechanism to offer developers more freedom and control over the implementation.

Your feedback and insights would be invaluable to ensuring this proposal aligns with our goals and addresses any potential challenges. Kindly review the RFC and share your thoughts.

Thank you.

### CODE REVIEW

File: rfcs/AP-RFC-003.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    - Add a URI parameter as an optional parameter
+    - Make file an optional parameter
+    - Require either URI or file to be specified
+    - Add an endpoint to the agent protocol to list the URI protocols that are valid for input and output for agents
+
+2. **Artifact Source Identification**:
+    - Introduce a mechanism, possibly via metadata, to discern if an artifact is generated by the agent or if it's an input.
+    
+3. **Revise Current Implementation**:
+    - Modify POST /agent/tasks/<task_id>/artifacts’s file parameter type from string to multipart/form-data
```
Comment (jakubno) [inline rfcs/AP-RFC-003.md:38]: What exactly do you mean by this?

File: rfcs/AP-RFC-003.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+## Design Proposal
+
+1. **Flexible Upload Paths**: 
+    - Allow agents to specify where files are uploaded from, accommodating external storage solutions such as AWS S3.
+    - Make relative path an optional parameter and mark it to be removed
+    - Add a URI parameter as an optional parameter
+    - Make file an optional parameter
+    - Require either URI or file to be specified
+    - Add an endpoint to the agent protocol to list the URI protocols that are valid for input and output for agents
```
Comment (jakubno) [inline rfcs/AP-RFC-003.md:32]: Something like `file://`, `https://`, `ftp://`, `s3`?

File: rfcs/AP-RFC-003.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  
+2. **Remove Artifact and Workspace Handling form Agent Class**:
+    - Remove the opinionated artifact and workspace handling logic from the agent class.
+
+3. **Require artifact upload and download handler functions**: 
+    - Require handler functions for artifact upload and download, which can be called form the protocol artifact endpoints.
+
+## Questions and Discussion Topics
+
+- Do we want to keep the artifact handling logic out of the agent protocol?
```
Comment (jakubno) [inline rfcs/AP-RFC-003.md:68]: I am more inclined to keep it out, only the interface should be given. Let's keep it unopinionated.

File: rfcs/AP-RFC-003.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+1. **Enhanced Artifact Class**:
+    - Retain the existing `Artifact` and `ArtifactUpload` classes.
+    - Incorporate attributes to identify artifact source (e.g., `sourceType` with possible values: 'input' or 'output').
+    - Provide methods/functions to specify external upload paths.
+  
+2. **Remove Artifact and Workspace Handling form Agent Class**:
+    - Remove the opinionated artifact and workspace handling logic from the agent class.
+
+3. **Require artifact upload and download handler functions**:
```
Comment (jakubno) [inline rfcs/AP-RFC-003.md:63]: This is implementation detail, maybe it's confusing. SDK in this repo is just our opinionated implementation, but it isn't part of standard. So it shouldn't probably be here.

Back to the point, I think it make sense to separate, so it's easily extensible or replaceable.

File: rfcs/AP-RFC-003.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+### Compatibility
+
+- Changes to the artifact class will only add additional fields with defaults set to preserve backward compatibility.
+- Removing the current artifact handling logic will break compatibility with existing agents.
+
+## Detailed Design (Optional)
+
+1. **Enhanced Artifact Class**:
+    - Retain the existing `Artifact` and `ArtifactUpload` classes.
+    - Incorporate attributes to identify artifact source (e.g., `sourceType` with possible values: 'input' or 'output').
```
Comment (jakubno) [inline rfcs/AP-RFC-003.md:57]: NIT: The names of the properties are now snake_case

```suggestion
    - Incorporate attributes to identify artifact source (e.g., `source_type` with possible values: 'input' or 'output').
```

File: rfcs/AP-RFC-003.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    - Make file an optional parameter
+    - Require either URI or file to be specified
+    - Add an endpoint to the agent protocol to list the URI protocols that are valid for input and output for agents
+
+2. **Artifact Source Identification**:
+    - Introduce a mechanism, possibly via metadata, to discern if an artifact is generated by the agent or if it's an input.
+    
+3. **Revise Current Implementation**:
+    - Modify POST /agent/tasks/<task_id>/artifacts’s file parameter type from string to multipart/form-data 
+    - Remove current artifact handling logic from agent class
```
Comment (jakubno) [inline rfcs/AP-RFC-003.md:39]: This isn't probably relevant to the protocol, but to SDK

**Review by jakubno — COMMENTED** [review]
Thanks a lot @Swiftyos! 
I really like this

File: rfcs/AP-RFC-003.md (hunk 4 above)
Comment (Swiftyos) [reply rfcs/AP-RFC-003.md:63]: Ah I understand I thought this RFC was for both

Comment (jakubno) [reply rfcs/AP-RFC-003.md:63]: No, RPC should be only about the protocol itself. You can create PR for SDK without any RPC, but it has to be conform to the protocol

**Comment by jakubno** [comment]
Could you please resolve comments and we can merge it

**Pull request closed without merging**